src/message_bar.rs

```diff
@@ -0,0 +1,476 @@
+                    line = split
+                } else {
+                    lines.push(Self::pad_text(line, num_cols));
+                    line = String::new();
+                }
+            }
+
+            if c != '\n' {
+                line.push(c);
+            }
+        }
+        lines.push(Self::pad_text(line, num_cols));
+
+        // Truncate output if it's too long
+        if lines.len() > max_lines {
+            lines.truncate(max_lines);
+            if TRUNCATED_MESSAGE.len() <= num_cols {
+                if let Some(line) = lines.iter_mut().last() {
+                    *line = Self::pad_text(TRUNCATED_MESSAGE.into(), num_cols);
+                }
+            }
+        }
+
+        // Append close button to first line
+        if CLOSE_BUTTON_TEXT.len() <= num_cols {
+            if let Some(line) = lines.get_mut(0) {
+                line.truncate(num_cols - CLOSE_BUTTON_TEXT.len());
+                line.push_str(CLOSE_BUTTON_TEXT);
+            }
+        }
+
+        lines
+    }
+
+    /// Message color
+    #[inline]
+    pub fn color(&self) -> Rgb {
+        self.color
+    }
+
+    /// Message topic
+    #[inline]
+    pub fn topic(&self) -> Option<&String> {
+        self.topic.as_ref()
+    }
+
+    /// Update the message topic
+    #[inline]
+    pub fn set_topic(&mut self, topic: String) {
+        self.topic = Some(topic);
+    }
+
+    /// Right-pad text to fit a specific number of columns
+    #[inline]
+    fn pad_text(mut text: String, num_cols: usize) -> String {
+        let padding_len = num_cols.saturating_sub(text.len());
+        text.extend(vec![' '; padding_len]);
+        text
+    }
+}
+
+/// Storage for message bar
+#[derive(Debug)]
+pub struct MessageBar {
+    current: Option<Message>,
+    messages: Receiver<Message>,
+    tx: Sender<Message>,
+}
+
+impl MessageBar {
+    /// Create new message bar
+    pub fn new() -> MessageBar {
+        let (tx, messages) = crossbeam_channel::unbounded();
+        MessageBar {
+            current: None,
+            messages,
+            tx,
+        }
+    }
+
+    /// Check if there are any messages queued
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.current.is_none()
+    }
+
+    /// Current message
+    #[inline]
+    pub fn message(&mut self) -> Option<Message> {
+        if let Some(current) = &self.current {
+            Some(current.clone())
+        } else {
+            self.current = self.messages.try_recv().ok();
+            self.current.clone()
+        }
+    }
+
+    /// Channel for adding new messages
+    #[inline]
+    pub fn tx(&self) -> Sender<Message> {
+        self.tx.clone()
+    }
+
+    /// Remove the currently visible message
+    #[inline]
+    pub fn pop(&mut self) -> Option<Message> {
+        std::mem::replace(&mut self.current, self.messages.try_recv().ok())
+    }
+
+    /// Remove all messages with a specific topic
+    #[inline]
+    pub fn remove_topic(&mut self, topic: String) {
```

Comment:
Since the `file!` return type is a `&'static str` anyways, I don't see much reason to make this a path. It would just add additional work to convert it to a path, not only a String.

If anything, I'd rather change this to a `&str`, so we wouldn't need to do the `&'static str` -> `String` conversion.

Reply:
Decided to change it to a `&str`!

Reply:
The only reason I mention this is that paths can be equivalent, but not the same string. Though Rust's `Path` doesn't resolve paths for equality checking anyway so it a moot point.

This looks good!